share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java

```diff
@@ -539,7 +539,7 @@ public void onNewMetadataImage(MetadataImage newImage, MetadataDelta delta) {
     }
 
     private TopicPartition topicPartitionFor(SharePartitionKey key) {
-        return new TopicPartition(Topic.SHARE_GROUP_STATE_TOPIC_NAME, partitionFor(key.toString()));
+        return new TopicPartition(Topic.SHARE_GROUP_STATE_TOPIC_NAME, partitionFor(key.asCoordinatorKey()));
```

Comment:
@AndrewJSchofield could we move away from using String keys to avoid issues like this in the future? IOW, could we refactor `partitionFor` to accept a SharePartitionKey?

Reply:
I created https://issues.apache.org/jira/browse/KAFKA-17914 to cover this. Unfortunately, it's a little more work than just changing this class, and I think it would be better to prioritise getting this merged.

Reply:
@mumrah Changing `partitionFor` would be counter productive, IMO. This is because it will be invoked by the KafkaApis FIND_COORD API handler. 

The find coordinator request comprises of the coordinator keys as Strings (as its a JSON defined RPC spec). The find coordinator handler will simply invoke `partitionFor` with the key specified in the request argument.

If we change `partitionFor` signature to accept the `SharePartitionKey`, we would not have enough information in FIND_COORD API handler to create a `SharePartitionKey` object and call `partitionFor` unless we resort to splitting and formatting string in the request and creating the required object and pass to `partitionFor`.

cc: @AndrewJSchofield

Reply:
@smjn I don't quite agree. I understand that the FindCoordinator RPC includes a string `group-id:topic-id:partition`. We could transform that into a `SharePartitionKey` on receipt, and then `ShareCoordinator.partitionFor` could take a `SharePartitionKey` as its parameter.

One other thing jumped out at me when I was debugging this. `SharePartitionKey` is a combination of a group ID and a `TopicIdPartition`. The topic name in the `TopicIdPartition` was null because it was constructed when no topic name was available. While a partition can be identified by just a topic ID and a partition, it does strike me that we need to be super-careful that we don't have a topic name sometimes, and no topic name at others, and introduce a bug due to this difference. One thing we could do for instance is ensure that the `TopicIdPartition` in a `SharePartitionKey` always has a null topic name. Maybe it doesn't make sense to use `TopicIdPartition.toString()` as part of the `SharePartitionKey.toString()` because the topic name is always null. Just wondering whether there's an easy win here in terms of defensive coding.

Reply:
@AndrewJSchofield 
> We could transform that into a SharePartitionKey on receipt
Which would mean writing code to split and format strings - was trying to avoid it.

> is a combination of a group ID and a TopicIdPartition
I did not want to do this. The initial impl of SPK clearly defined 3 fields -String groupId, Uuid topicId, int partition. In fact, the entire share-coordinator flow does not use the TopicIdPartition argument to the constructor at all. Not sure whether it is necessary for the SharePartition flow or was it done for aesthetic reasons.

Reply:
> the entire share-coordinator flow does not use the TopicIdPartition argument to the constructor at all

IMO, we should use TopicIdPartition class itself as used elsewhere, not beacause of aesthetic reasons but it's a logical encapsulated entity.

Reply:
>  but it's a logical encapsulated entity.

Yes thats true but it might not be a good fit since topicName is always  (or mostly) null.

Reply:
> Yes thats true but it might not be a good fit since topicName is always (or mostly) null.

And why is that? For the calls from SharePartition we can fill up topicName as well. We do a metadata read to verify first.

Reply:
If you plan to use it then it should be fine. Otherwise, serves no purpose.